### PR TITLE
feat(app)!: cosmos SDK v0.46 upgrade

### PR DESCRIPTION
**Pull request by aleem1314**
## Description

Closes: #857

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] followed the guidelines for [building modules](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules)
- [ ] included the necessary unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed 
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

**Comment by ryanchristo** [comment]
If v3 ecocredit migrations are a blocker because of the orm module dependency on sdk version v0.45, I would say uncomment or remove the migrations for now and maybe we can further discuss on our next architecture call. The extra work to support v3 to v5 or later seems unnecessary but it is the right idea.

I think given we are not aware of any other chains running v3, our ecocredit api was tagged v1alpha1, and sdk versioning is the main issue here, we don't need to go through the extra trouble to support upgrades directly from v3 to v5 or later.

File: go.mod [hunk 1]
```diff
@@ -48,98 +57,111 @@ require (
...
+	gopkg.in/yaml.v3 v3.0.1 // indirect
 	nhooyr.io/websocket v1.8.6 // indirect
+	sigs.k8s.io/yaml v1.3.0 // indirect
 )
 
-replace google.golang.org/grpc => google.golang.org/grpc v1.33.2
-
 replace github.com/gogo/protobuf => github.com/regen-network/protobuf v1.3.3-alpha.regen.1
 
-replace github.com/cosmos/cosmos-sdk => github.com/regen-network/cosmos-sdk v0.45.0-regen-1
```
Comment (aleem1314) [inline go.mod:141]: Removing regen fork. Permanent locked vesting accounts is up streamed to SDK and it is part of v0.46.

File: Makefile [hunk 2]
```diff
@@ -92,7 +92,8 @@ ldflags = -X github.com/cosmos/cosmos-sdk/version.Name=regen \
 	-X github.com/cosmos/cosmos-sdk/version.AppName=regen \
 	-X github.com/cosmos/cosmos-sdk/version.Version=$(VERSION) \
 	-X github.com/cosmos/cosmos-sdk/version.Commit=$(COMMIT) \
-	-X github.com/cosmos/cosmos-sdk/version.BuildTags=$(build_tags_comma_sep)
+	-X github.com/cosmos/cosmos-sdk/version.BuildTags=$(build_tags_comma_sep) \
+	-X google.golang.org/protobuf/reflect/protoregistry.conflictPolicy=warn
```
Comment (aleem1314) [inline Makefile:96]: I'm getting proto namespace-conflict error, not sure how to fix this.

```
WARNING: proto: file "cosmos/base/query/v1beta1/pagination.proto" is already registered
        previously from: "cosmossdk.io/api/cosmos/base/query/v1beta1"
        currently from:  "github.com/cosmos/cosmos-sdk/api/cosmos/base/query/v1beta1"
See https://developers.google.com/protocol-buffers/docs/reference/go/faq#namespace-conflict

WARNING: proto: file "cosmos/base/query/v1beta1/pagination.proto" has a name conflict over cosmos.base.query.v1beta1.PageResponse
        previously from: "cosmossdk.io/api/cosmos/base/query/v1beta1"
        currently from:  "github.com/cosmos/cosmos-sdk/api/cosmos/base/query/v1beta1"
See https://developers.google.com/protocol-buffers/docs/reference/go/faq#namespace-conflict

WARNING: proto: file "cosmos/base/query/v1beta1/pagination.proto" has a name conflict over cosmos.base.query.v1beta1.PageRequest
        previously from: "cosmossdk.io/api/cosmos/base/query/v1beta1"
        currently from:  "github.com/cosmos/cosmos-sdk/api/cosmos/base/query/v1beta1"
See https://developers.google.com/protocol-buffers/docs/reference/go/faq#namespace-conflict

WARNING: proto: message cosmos.base.query.v1beta1.PageRequest is already registered
        previously from: "cosmossdk.io/api/cosmos/base/query/v1beta1"
        currently from:  "github.com/cosmos/cosmos-sdk/api/cosmos/base/query/v1beta1"
See https://developers.google.com/protocol-buffers/docs/reference/go/faq#namespace-conflict

WARNING: proto: message cosmos.base.query.v1beta1.PageResponse is already registered
        previously from: "cosmossdk.io/api/cosmos/base/query/v1beta1"
        currently from:  "github.com/cosmos/cosmos-sdk/api/cosmos/base/query/v1beta1"
See https://developers.google.com/protocol-buffers/docs/reference/go/faq#namespace-conflict

WARNING: proto: file "cosmos/base/v1beta1/coin.proto" is already registered
        previously from: "cosmossdk.io/api/cosmos/base/v1beta1"
        currently from:  "github.com/cosmos/cosmos-sdk/api/cosmos/base/v1beta1"
See https://developers.google.com/protocol-buffers/docs/reference/go/faq#namespace-conflict

WARNING: proto: file "cosmos/base/v1beta1/coin.proto" has a name conflict over cosmos.base.v1beta1.DecProto
        previously from: "cosmossdk.io/api/cosmos/base/v1beta1"
        currently from:  "github.com/cosmos/cosmos-sdk/api/cosmos/base/v1beta1"
See https://developers.google.com/protocol-buffers/docs/reference/go/faq#namespace-conflict

WARNING: proto: file "cosmos/base/v1beta1/coin.proto" has a name conflict over cosmos.base.v1beta1.IntProto
        previously from: "cosmossdk.io/api/cosmos/base/v1beta1"
        currently from:  "github.com/cosmos/cosmos-sdk/api/cosmos/base/v1beta1"
See https://developers.google.com/protocol-buffers/docs/reference/go/faq#namespace-conflict

WARNING: proto: file "cosmos/base/v1beta1/coin.proto" has a name conflict over cosmos.base.v1beta1.DecCoin
        previously from: "cosmossdk.io/api/cosmos/base/v1beta1"
        currently from:  "github.com/cosmos/cosmos-sdk/api/cosmos/base/v1beta1"
See https://developers.google.com/protocol-buffers/docs/reference/go/faq#namespace-conflict

WARNING: proto: file "cosmos/base/v1beta1/coin.proto" has a name conflict over cosmos.base.v1beta1.Coin
        previously from: "cosmossdk.io/api/cosmos/base/v1beta1"
        currently from:  "github.com/cosmos/cosmos-sdk/api/cosmos/base/v1beta1"
See https://developers.google.com/protocol-buffers/docs/reference/go/faq#namespace-conflict

WARNING: proto: message cosmos.base.v1beta1.Coin is already registered
        previously from: "cosmossdk.io/api/cosmos/base/v1beta1"
        currently from:  "github.com/cosmos/cosmos-sdk/api/cosmos/base/v1beta1"
See https://developers.google.com/protocol-buffers/docs/reference/go/faq#namespace-conflict

WARNING: proto: message cosmos.base.v1beta1.DecCoin is already registered
        previously from: "cosmossdk.io/api/cosmos/base/v1beta1"
        currently from:  "github.com/cosmos/cosmos-sdk/api/cosmos/base/v1beta1"
See https://developers.google.com/protocol-buffers/docs/reference/go/faq#namespace-conflict

WARNING: proto: message cosmos.base.v1beta1.IntProto is already registered
        previously from: "cosmossdk.io/api/cosmos/base/v1beta1"
        currently from:  "github.com/cosmos/cosmos-sdk/api/cosmos/base/v1beta1"
See https://developers.google.com/protocol-buffers/docs/reference/go/faq#namespace-conflict

WARNING: proto: message cosmos.base.v1beta1.DecProto is already registered
        previously from: "cosmossdk.io/api/cosmos/base/v1beta1"
        currently from:  "github.com/cosmos/cosmos-sdk/api/cosmos/base/v1beta1"
See https://developers.google.com/protocol-buffers/docs/reference/go/faq#namespace-conflict

```

Comment (aleem1314) [reply Makefile:96]: Looks like issue is from the SDK side. I think we should remove [cosmossdk.io/api](https://github.com/cosmos/cosmos-sdk/blob/v0.46.0-rc3/go.mod#L6) import from the `go.mod`.

Comment (aleem1314) [reply Makefile]: resolved by https://github.com/cosmos/cosmos-sdk/pull/12719

**Comment by aleem1314** [comment]
Marking as R4R.

**Comment by technicallyty** [comment]
> option (cosmos.msg.v1.signer)

only thing i could possibly think of is a missing import? `import "cosmos/msg/v1/msg.proto";` i know robert had a similar issue in the cosmos core dev discord: https://discord.com/channels/669268347736686612/984437601857568768/1002138288250957834

File: x/data/msg_anchor.go [hunk 3]
```diff
@@ -3,7 +3,7 @@ package data
 import (
 	sdk "github.com/cosmos/cosmos-sdk/types"
 	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
-	"github.com/cosmos/cosmos-sdk/x/auth/legacy/legacytx"
+	"github.com/cosmos/cosmos-sdk/x/auth/migrations/legacytx"
```
Comment (ryanchristo) [inline x/data/msg_anchor.go:6]: Ok, to leave for now but wondering if we need to still support legacy messages...?

**Review by ryanchristo — COMMENTED** [review]
Nice work! Looks good to me aside from a few minor changes, in particular the passing of the account prefix rather than the bond denom to the account keeper.

File: x/data/msg_anchor.go (hunk 3 above)
Comment (aleem1314) [reply x/data/msg_anchor.go:6]: Yes, for ledger signing.

File: x/ecocredit/server/tests/utils.go [hunk 4]
```diff
@@ -48,7 +49,7 @@ func NewEcocreditModule(ff *server.FixtureFactory) *ecocredit.Module {
 	}
 
 	accountKeeper := authkeeper.NewAccountKeeper(
-		cdc, authKey, authSubspace, authtypes.ProtoBaseAccount, maccPerms,
+		cdc, authKey, authSubspace, authtypes.ProtoBaseAccount, maccPerms, "regen",
```
[the remark this answers is not in the document]
Comment (aleem1314) [reply x/ecocredit/server/tests/utils.go:52]: Importing the app module in ecocredit creates an import cycle. We are only using `Bech32PrefixAccAddr` in tests. I think it is okay to hardcode.

File: scripts/test_cover.sh [hunk 5]
```diff
@@ -15,7 +15,7 @@ for m in ${SUBMODULES[@]}; do
     cd $CURDIR/$m
     PKGS=$(go list ./...)
     for pkg in ${PKGS[@]}; do
-        go test -v -timeout 30m -race -coverpkg=all -coverprofile=profile.out -covermode=atomic -tags="ledger test_ledger_mock $EXPERIMENTAL_TAG" "$pkg"
+        go test -v -timeout 30m -coverpkg=all -coverprofile=profile.out -covermode=atomic -tags="norace ledger test_ledger_mock rocksdb_build $EXPERIMENTAL_TAG" "$pkg"
```
[the remark this answers is not in the document]
Comment (aleem1314) [reply scripts/test_cover.sh]: I copied it from SDK. Switched it back to default db.

File: app/app.go [hunk 6]
```diff
@@ -298,6 +278,7 @@ func NewRegenApp(logger log.Logger, db dbm.DB, traceStore io.Writer, loadLatest
 	// add keepers
 	app.AccountKeeper = authkeeper.NewAccountKeeper(
 		appCodec, keys[authtypes.StoreKey], app.GetSubspace(authtypes.ModuleName), authtypes.ProtoBaseAccount, maccPerms,
+		"uregen",
```
[the remark this answers is not in the document]
Comment (ryanchristo) [reply app/app.go:275]: Looks like this was resolved without being addressed. This should be the bech32 account prefix.

**Review by ryanchristo — APPROVED** [review]
Nice work! One item that still needs to be addressed in `app.go` (we are still using `uregen` rather than the bech32 account prefix). Otherwise, looks good to me. I opened followup issues for the two items you mentioned above: #1330 and #1331.

**Pull request merged**